Add `.d.ts` files for improved TypeScript support

patricknelson opened the pull request:
Implement #30.

patricknelson commented:
@baseplate-admin can you test this out for me please? Best method is going to be for you to clone locally and link it in your project. Here some quick guidance though

1. Clone `svelte-retag` to your system
4. Run `yarn link` somewhere inside the `svelte-retag` folder
5. Go to your project and install `svelte-retag` (if not already)
6. Run `yarn link svelte-retag` to symbolically link the project so you can edit it in-place from the originally cloned dir

Once done testing, reverse the link to your local copy via `yarn unlink svelte-retag` and then `npm i` (or `yarn` or whatever your local package manager is). I don't have a formal contribution guidelines doc yet that explains how to clone and run locally but hopefully you get the gist and I'm just overexplaining 😄 (so, sorry if this messes something up)

baseplate-admin commented on .gitignore:
```diff
@@ -5,3 +5,6 @@ coverage
 # In case contributors use yarn (just prevent it from being committed). Standardizing on npm for broad compatibility.
 yarn.lock
+
+# Ignore autogenerated .d.ts (and associated .map) files since this are automatically generated prior to publish to NPM.
```
You shouldn’t ignore this file. Most repos commit this file

patricknelson replied:
Would the expectation w/ [vite-plugin-dts](https://www.npmjs.com/package/vite-plugin-dts) be to run that during development and to also commit changes to those `.d.ts` files? Right now I'm still not convinced we should commit those. 

Again, the reason why is because I do not want to incorporate that into every PR, every commit and etc. The dev workflow is far easier right now keeping it with JSDoc and only instead pushing out the `.d.ts` on publish for developer convenience (in case they are in your situation where JSDoc for some reason doesn't work in your IDE). 

For example, in Jetbrains (PhpStorm is what I'm using), the JSDoc types are pulled in just fine. It's possible this is an issue with VSCode or maybe more specifically your extensions in VSCode, I'm not sure. 

Finally:

> It's easy to directly link issues in typescript hint files, instead of going through the codebase to find out where it was declared.

This may just be a difference of opinion. This is actually what happens for me _**if**_ the `.d.ts` are present; it requires hunting because you get linked to the `.d.ts` file instead of the `.js` file. This is possibly an argument to _not_ include them in the published packages, since 99% of folks using it will not be using a cloned/linked version but instead one with the combined `.js` paired with the `.d.js` files.

patricknelson commented:
p.s. Since I didn’t mention it explicitly (sorry I forgot): Be sure to run `npm run build` to generate those `.d.ts` files in the root of the main `svelte-retag` folder.

patricknelson commented:
p.s. @baseplate-admin in my main TypeScript project, this is what I did. However, I am not 100% sure what would need to be done in a JS-only project like this one when using `.d.ts` files instead. So, your PR would be helpful (which is partly why I ask). 

Just to give you an idea, here what I ended up doing (probably pretty amateur):

```typescript
// Generic type for anything that can be instantiated.  Copied and modified from https://github.com/microsoft/TypeScript/issues/17572#issue-247539885
export type Instantiable = {new(...args: any[]): any};

// Setup to type hint Svelte component constructors (i.e. the default export of *.svelte files).
export type SvelteComponent = Instantiable;
```

Then where I use `svelteRetag` I just pull that in, e.g. 

```typescript
import svelteRetag from 'svelte-retag';
import type { SvelteComponent } from './types';

/**
 * Defines a custom element for the given Svelte component and hyphenated tag name.
 *
 * @param {SvelteComponent}	component The svelte component constructor. Basically, the default export from your MyComponent.svelte file.
 *
 * @param {string}          tagName	  The name of the tag. Must contain a dash and be all lower case.
 *                                    e.g. 'my-component' (for <my-component> tag).
 *
 * @param {array}           syncProps For reactively synchronizing changes to the element attributes with the props on
 *                                    the Svelte component. Not necessary for initialization, just for changes
 *                                    performed after mount.
 */
export function defineTag(component: SvelteComponent, tagName: string, syncProps = []) {
	svelteRetag({
		component: component,
		tagname: tagName,
		attributes: syncProps,
	});
}
```

baseplate-admin commented:
Fair enough, give me some time to get personal life sorted then i will send a PR

baseplate-admin commented:
Hi, Sorry for late reply. This works fine for my project :) 

Awesome work. Can we get this merged?

patricknelson commented:
Fair warning @baseplate-admin `main` has changed a lot, so I'm rebasing, so you'll need to do a hard reset (or also rebase) your local version if you're still testing it locally. Also just to summarize for you:

1. Rebased (see above)
2. Updated type definitions and using an "instantiable" (or "newable") type to define the component constructor, which is what we're taking (not a component instance per se).
3. **New:** I forgot to mention that, in this branch, I fixed type definitions so that you don't see errors for including _**optional**_ flags like `href` or `hydratable`. 
4. **New:** I added automatic attribute forwarding in `svelte-retag@1.4.0` which means you no longer need to define every damned attribute. 😅 Give that a whirl (just set `attributes: true`). This means you can use `import.meta.glob()` to just automatically define everything without having to specify individual attributes. If you are feeling _risky_, check out the example `autoDefine` function I setup here https://github.com/patricknelson/svelte-retag/blob/main/demo/vercel/src/main.js#L39-L88 (but use at your own risk, I'd suggest just modifying it to suit your needs for now). However, it does take an `options` callback as a third parameter which can [accept `tagName` and `options` object](https://github.com/patricknelson/svelte-retag/blob/main/demo/vercel/src/main.js#L73) that you can adjust as needed, if needed.

Good luck.

patricknelson commented:
You can also test this as a beta release, too (what I'm doing in my project, since symlinks aren't great for Intellisense in my case on Windows):

```bash
npm install svelte-retag@1.5.0-beta.0
```

baseplate-admin commented:
```typescript
import svelteRetag from "svelte-retag";

const mappings = [
    { tagname: "markdown", component: await import("$components/minor/Markdown/Index.svelte"), attributes: [`markdown`, `class`] },
    { tagname: "scroll-area", component: await import("$components/minor/ScrollArea/Index.svelte"), attributes: [`parent_class`, `offset_scrollbar`, `gradient_mask`, `class`] },
    { tagname: "comment", component: await import("$components/minor/Comment/Index.svelte"), attributes: [`api_url`] }
];

mappings.forEach((item) => {
    svelteRetag({
        component: item.component.default,
        tagname: `coreproject-${item.tagname}`,
        attributes: item.attributes, // Changes to these attributes will be reactively forwarded to your component
        shadow: false, // Use the light DOM
        hydratable: false
    });
});

```

This type of code works with `svelte-retag@1.5.0-beta.0`

patricknelson commented:
Sweet. I’ll work on this tomorrow and probably have 1.5.0 out by then.

Also @baseplate-admin

1. When 1.5.0 is out, you won’t need to set `shadow` or `hydratable` in TypeScript since this branch also fixes the issue of those options showing up as required.
2. Have you considered grouping your components into folders and then using `import.meta.glob` to automate (without mapping)? That should now be feasible in 1.4.0 w/ the `attributes: true` option since you no longer need to cater each call to `svelteRetag()` to include the customized attribute list. 😊

patricknelson commented on index.js:
```diff
@@ -66,24 +66,29 @@ function renderElements(timestamp) {
 
+/**
+ * @typedef {new(...args: any[]) => any} Newable         Type alias for a really generic class constructor
+ * @typedef {Newable}                    SvelteComponent Svelte component class constructor (basically a "newable" object)
```
Consider rename, e.g. maybe `ComponentType` (https://svelte.dev/docs/typescript#types-componenttype) or make own like `ComponentConstructor` for clarity, since in Svelte, `SvelteComponent` actually exists and is an instance type… I think.

If `ComponentType` is used though, it should probably be actually imported.

patricknelson commented:
This is published now as `svelte-retag@1.5.0` but noticing some issues in PhpStorm right now. Might work fine in VSCode though. Still need to figure out why the types aren't being properly inferred in my IDE. 🤕

patricknelson commented:
Ah, never mind... false alarm. It was a caching issue on my end.
